docs/root/development/performance/binary_size.rst

```diff
@@ -0,0 +1,164 @@
+instance with Ubuntu 18.04. The following script was run to set up the
+necessary tools::
+
+  # basic toolchain
+  sudo apt-get install build-essential openjdk-8-jdk python zip unzip \
+    software-properties-common make cmake bc libtool ninja-build automake \
+    time apt-transport-https
+  # clang-8
+  wget http://releases.llvm.org/8.0.0/clang+llvm-8.0.0-aarch64-linux-gnu.tar.xz
+  sudo tar -C /usr/local/ -xvf clang+llvm-8.0.0-aarch64-linux-gnu.tar.xz --strip 1
+  rm -rf clang+llvm-8.0.0-aarch64-linux-gnu.tar.xz
+  # bazel 0.26.1
+  wget https://github.com/bazelbuild/bazel/releases/download/0.26.1/bazel-0.26.1-dist.zip
+  mkdir -p /tmp/bazel_build
+  unzip -o bazel-0.26.1-dist.zip -d /tmp/bazel_build
+  rm -rf bazel-0.26.1-dist.zip
+  cd /tmp/bazel_build
+  env EXTRA_BAZEL_ARGS="--host_javabase=@local_jdk//:jdk" bash ./compile.sh
+  sudo cp output/bazel /usr/local/bin/bazel
+  rm -rf /tmp/bazel_build
+  # bloaty
+  git clone https://github.com/google/bloaty
+  cd bloaty
+  mkdir build
+  cd build
+  cmake ..
+  make -j6
+  cp bloaty /usr/local/bin/bloaty
+
+The binary being compiled is ``//test/performance:test_binary_size``.
+The binary is getting built with the following build command::
+
+  bazel build //test/performance:test_binary_size --config=sizeopt --copt=-ggdb3 --linkopt=-fuse-ld=lld
+
+Thus the binary is compiled with the following flags:
+
+.. _envoy_docs: https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#enabling-optional-features
+
+1. ``-c opt``: bazel compilation option for size optimization.
+2. ``--copt -Os``: optimize for size.
+3. ``--copt=-ggdb3``: keep debug symbols. Later stripped with ``strip``
+4. ``--linkopt=-fuse-ld=lld``: use the lld linker.
+5. ``--define=google_grpc=disabled``: more info in the `envoy docs <envoy_docs>`_.
+6. ``--define=signal_trace=disabled``: more info in the `envoy docs <envoy_docs>`_.
+7. ``--define=tcmalloc=disabled``: more info in the `envoy docs <envoy_docs>`_.
+8. ``--define=hot_restart=disabled``: more info in the `envoy docs <envoy_docs>`_.
+
+After compiling, the binary can be stripped of all symbols by using ``strip``::
+
+  strip -s bazel-bin/test/performance/test_binary_size
+
+The unstripped and stripped binary can then be used for analysis.
+
+Analysis
+~~~~~~~~
+
+While there are a lot of tools out there that can be used for binary size
+analysis (otool, objdump, jtool), `Bloaty <https://github.com/google/bloaty>`_
+has been the tool of choice to run object file analysis.
+
+Bloaty's layering of data sources is extremely helpful in being able to explode
+the binary in all sorts of different ways. For example, one can look at the
+composition of each compile unit in terms of sections::
+
+  $ bloaty --debug-file=bin/test_binary_size -c envoy.bloaty -d sections,bloaty_package,compileunits bin/test_binary_size.stripped
+  ...
+  7.7%   109Ki   7.7%   110Ki bazel-out/aarch64-opt/bin/external/envoy_api/envoy/api/v2/route/route.pb.cc
```

Comment:
Is this an intentional format choice?

```suggestion
    7.7%   109Ki   7.7%   110Ki bazel-out/aarch64-opt/bin/external/envoy_api/envoy/api/v2/route/route.pb.cc
```

Reply:
yep that is bloaty's indentation to show groups of data.